### ISSUE TITLE
Rotas de Fuga rev 5

### ISSUE THREAD

**Issue by guilhermegandra** [guilhermegandra/Doutorado#5]
print("""
Bem-vindo ao Sistema Automatizado de Geração de Rotas de Fuga!

Este algoritmo foi desenvolvido para auxiliar equipes técnicas, pesquisadores e gestores de risco na criação rápida e precisa de rotas de fuga baseadas em rede viária, considerando cenários de inundação e a distribuição espacial das edificações afetadas.

Para que tudo funcione corretamente, você precisará fornecer arquivos compactados em formato ZIP, cada um contendo um shapefile completo (arquivos .shp, .shx, .dbf, .prj) ou, opcionalmente, um Modelo Digital do Terreno (MDT) em formato raster (GeoTIFF, também compactado).

Os dados devem obedecer à seguinte topologia:
• Rede Viária (polilinhas);
• Edificações ou Imóveis (pontos);
• Mancha de Inundação (polígono);
• MDT (opcional).

É fundamental que **todos** os dados espaciais estejam no sistema de referência UTM correspondente ao fuso da área de estudo.

Após o carregamento dos dados, o algoritmo irá:
1. Corrigir automaticamente a rede viária por meio de operações de snap e split, garantindo conectividade topológica adequada para roteamento;
2. Identificar automaticamente todas as edificações situadas dentro da mancha de inundação;
3. Solicitar ao usuário:
   • A velocidade média de deslocamento da população (km/h);
   • O tempo máximo admissível de deslocamento até um ponto de encontro;
   • A existência de um MDT para cálculo de impedâncias associadas à declividade do terreno;
   • A existência de pontos de encontro previamente definidos;
4. Construir um grafo viário ponderado pelo tempo de deslocamento, incorporando, quando disponível, penalizações topográficas baseadas na declividade;
5. Calcular a rota mais eficiente entre cada edificação afetada e seu respectivo ponto de encontro;
6. Caso alguma edificação não consiga atingir um ponto de encontro dentro do tempo máximo estabelecido, o algoritmo identificará automaticamente novos pontos de encontro por meio de agrupamento espacial;
7. Caso o usuário forneça pontos de encontro válidos, estes poderão ser utilizados como destinos preferenciais, sem impedir a criação automática de pontos adicionais quando necessário.

Ao final do processamento, o sistema exportará automaticamente:
• Rotas individuais em formato SHP e KML para cada edificação afetada;
• Rotas consolidadas;
• Pontos de encontro utilizados;
• Edificações afetadas com atributos de tempo, distância e declividade média;
• Relatório técnico detalhado em formato TXT, contendo auditoria da rede, estatísticas do grafo, tempos de processamento e eventuais inconsistências;
• Relatório analítico em formato XLSX;
• Um arquivo GeoPackage consolidando todos os resultados;
• Um arquivo ZIP contendo toda a estrutura de saída organizada.

Bons trabalhos e uma excelente análise!
""")

!apt-get install -y libspatialindex-dev > /dev/null 2>&1
!pip install -q geopandas shapely fiona rtree pyproj tqdm networkx scipy scikit-learn openpyxl pandas rasterio

import os, zipfile, math, time, traceback
import pandas as pd
import numpy as np
import geopandas as gpd
import networkx as nx
from shapely.geometry import Point, LineString
from shapely.ops import snap, split, unary_union, nearest_points
from tqdm import tqdm
from scipy.spatial import cKDTree
from sklearn.cluster import DBSCAN
from google.colab import files
import rasterio

log_lines = []
log_errors = []

def log(msg):
    print(msg)
    log_lines.append(str(msg))

def log_error(msg):
    print(msg)
    log_errors.append(str(msg))

def start_stage(nome):
    log(f"ETAPA: {nome}")
    return time.perf_counter(), nome

def end_stage(s):
    t0, nome = s
    log(f"Tempo da etapa {nome}: {time.perf_counter() - t0:.2f} s")

def upload_and_extract(label):
    log(f"Envie o ZIP contendo o shapefile de: {label}")
    uploaded = files.upload()
    zip_name = list(uploaded.keys())[0]
    out = f"input_{label}"
    os.makedirs(out, exist_ok=True)
    with zipfile.ZipFile(zip_name) as z:
        z.extractall(out)
    shp = [f for f in os.listdir(out) if f.lower().endswith(".shp")]
    if not shp:
        raise ValueError(f"O ZIP de {label} não contém shapefile")
    return os.path.join(out, shp[0])

def upload_mdt():
    log("Carregue o ZIP contendo o MDT (GeoTIFF)")
    uploaded = files.upload()
    zip_name = list(uploaded.keys())[0]
    out = "input_MDT"
    os.makedirs(out, exist_ok=True)
    with zipfile.ZipFile(zip_name) as z:
        z.extractall(out)
    tifs = [f for f in os.listdir(out) if f.lower().endswith((".tif", ".tiff"))]
    if not tifs:
        raise ValueError("ZIP do MDT inválido")
    return os.path.join(out, tifs[0])

def garantir_crs(gdf, nome):
    if gdf.crs is None:
        epsg = int(input(f"{nome} sem CRS. Informe EPSG: "))
        gdf = gdf.set_crs(epsg)
    return gdf

def tobler_speed_kmh(g):
    return 6.0 * math.exp(-3.5 * abs(g + 0.05))

def kmh_to_ms(v):
    return v * 1000 / 3600

def nearest_node(pt, tree, nodes):
    _, idx = tree.query([pt.x, pt.y])
    return tuple(nodes[idx])

VSEG_MIN_KMH = 0.3
SLOPE_CLAMP = 1.0

try:
    s = start_stage("Upload dos dados")
    vias_path = upload_and_extract("VIAS")
    edific_path = upload_and_extract("EDIFICACOES")
    mancha_path = upload_and_extract("MANCHA")
    end_stage(s)

    s = start_stage("Leitura")
    vias = gpd.read_file(vias_path)
    edific = gpd.read_file(edific_path)
    mancha = gpd.read_file(mancha_path)
    end_stage(s)

    vias = garantir_crs(vias, "vias").to_crs(31982)
    edific = garantir_crs(edific, "edificacoes").to_crs(31982)
    mancha = garantir_crs(mancha, "mancha").to_crs(31982)
    flood_union = mancha.geometry.unary_union

    vel_kmh = float(input("Velocidade média de deslocamento (km/h): "))
    tempo_max = float(input("Tempo máximo permitido (min): "))
    vel_ms = kmh_to_ms(vel_kmh)

    s = start_stage("Correção da rede viária")
    vias = vias[~vias.geometry.is_empty & ~vias.geometry.isna()]
    linhas = []
    for g in vias.geometry:
        linhas.extend(g.geoms if g.geom_type == "MultiLineString" else [g])

    pts = [Point(c[:2]) for l in linhas for c in l.coords]
    union_pts = gpd.GeoSeries(pts, crs=vias.crs).unary_union
    snapped = [snap(l, union_pts, 0.5) for l in linhas]

    vias_split = []
    for i in range(len(snapped)):
        sl = snapped[i]
        for j in range(len(snapped)):
            if i != j and sl.intersects(snapped[j]):
                try:
                    sl = split(sl, snapped[j])
                except:
                    pass
        if sl.geom_type == "MultiLineString":
            vias_split.extend(sl.geoms)
        else:
            vias_split.append(sl)
    end_stage(s)

    usa_mdt = input("A área analisada possui MDT? (Sim/Nao): ").lower()
    has_mdt = False
    if usa_mdt == "sim":
        mdt = rasterio.open(upload_mdt())
        has_mdt = True

    s = start_stage("Construção do grafo")
    G = nx.Graph()
    node_alt = {}

    for l in vias_split:
        for c in l.coords:
            node_alt[c] = np.nan
            if has_mdt:
                try:
                    for v in mdt.sample([(c[0], c[1])]):
                        node_alt[c] = float(v[0])
                except:
                    pass

        for a, b in zip(l.coords[:-1], l.coords[1:]):
            d = math.dist(a, b)
            if d <= 0:
                continue
            t = (d / vel_ms) / 60
            if has_mdt:
                ha, hb = node_alt[a], node_alt[b]
                if not np.isnan(ha) and not np.isnan(hb):
                    g = (hb - ha) / d
                    g = max(min(g, SLOPE_CLAMP), -SLOPE_CLAMP)
                    vseg = min(tobler_speed_kmh(g), vel_kmh)
                    vseg = max(vseg, VSEG_MIN_KMH)
                    t = (d / kmh_to_ms(vseg)) / 60
            G.add_edge(a, b, weight=max(t, 0.0001))

    nodes = np.array(list(G.nodes))
    tree = cKDTree(nodes)
    end_stage(s)

    s = start_stage("Identificação de edificações afetadas")
    afetadas = gpd.sjoin(edific, mancha, how="inner", predicate="intersects").reset_index(drop=True)
    end_stage(s)

    s = start_stage("Roteamento")
    rotas, tempos = [], []
    for _, row in afetadas.iterrows():
        try:
            start = nearest_node(row.geometry.centroid, tree, nodes)
            centro = row.geometry.centroid
            if centro.within(flood_union):
                centro = nearest_points(centro, flood_union.boundary)[1]
            dest = nearest_node(centro, tree, nodes)
            t = math.ceil(nx.shortest_path_length(G, start, dest, weight="weight"))
            p = nx.shortest_path(G, start, dest, weight="weight")
            rotas.append(LineString(p))
            tempos.append(t)
        except:
            rotas.append(None)
            tempos.append(None)
    end_stage(s)

    s = start_stage("Exportação")
    os.makedirs("saida", exist_ok=True)
    gpd.GeoDataFrame({"tempo_min": tempos}, geometry=rotas, crs=vias.crs)\
        .dropna().to_file("saida/rotas_consolidadas.shp")
    afetadas["tempo_min"] = tempos
    afetadas.to_file("saida/edificacoes_afetadas.shp")

    with zipfile.ZipFile("resultados_rotas_fuga_REV07_1.zip", "w", zipfile.ZIP_DEFLATED) as z:
        for root, _, files_ in os.walk("saida"):
            for f in files_:
                z.write(os.path.join(root, f), arcname=f)
    files.download("resultados_rotas_fuga_REV07_1.zip")
    end_stage(s)

    print("Processamento finalizado com sucesso.")

except Exception as e:
    log_error(str(e))
    log_error(traceback.format_exc())
    print("Erro durante a execução.")
